### PR TITLE
NA Tangent reverse implication

### PR DESCRIPTION
**Pull request by ahmed-irfan**
This PR adds additional tangent plane lemmas (reverse implication: FROM the comparison between a tangent plane and a multiplication TO bounds on the multiplicands).

We have run a small test on QF_NIA benchmarks with 60sec timeout. (Thanks to @yoni206 )
These additional lemmas helped to solve about 250 more benchmarks. 

image

### CODE REVIEW

**Comment by ajreynol** [comment]
This is excellent :)

File: src/theory/arith/nonlinear_extension.cpp [hunk 1]
```diff
@@ -3360,6 +3360,53 @@ std::vector<Node> NonlinearExtension::checkTangentPlanes() {
...
+                    NodeManager::currentNM()->mkNode(OR, b_geq_bv, a_geq_av);
+                Node ub_reverse1 = NodeManager::currentNM()->mkNode(
+                    OR, t_leq_tplane.negate(), a_leq_av_or_b_leq_bv);
+                Trace("nl-ext-tplanes")
+                    << "Tangent plane lemma : " << ub_reverse1 << std::endl;
+                lemmas.push_back(ub_reverse1);
+                Node ub_reverse2 = NodeManager::currentNM()->mkNode(
+                    OR, t_leq_tplane.negate(), b_geq_bv_or_a_geq_av);
+                Trace("nl-ext-tplanes")
+                    << "Tangent plane lemma : " << ub_reverse2 << std::endl;
```
Comment (ajreynol) [inline src/theory/arith/nonlinear_extension.cpp]: Can you change this trace message to something that distinguishes it, like "Tangent plane lemma (reverse) : ".

File: src/theory/arith/nonlinear_extension.cpp [hunk 2]
```diff
@@ -3360,6 +3360,53 @@ std::vector<Node> NonlinearExtension::checkTangentPlanes() {
...
+                Node t_geq_tplane =
+                    NodeManager::currentNM()->mkNode(GEQ, t, tplane);
+                Node a_leq_av_or_b_geq_bv =
+                    NodeManager::currentNM()->mkNode(OR, a_leq_av, b_geq_bv);
+                Node a_geq_av_or_b_leq_bv =
+                    NodeManager::currentNM()->mkNode(OR, a_geq_av, b_leq_bv);
+                Node lb_reverse1 = NodeManager::currentNM()->mkNode(
+                    OR, t_geq_tplane.negate(), a_leq_av_or_b_geq_bv);
+                Trace("nl-ext-tplanes")
+                    << "Tangent plane lemma : " << lb_reverse1 << std::endl;
```
Comment (ajreynol) [inline src/theory/arith/nonlinear_extension.cpp]: Ditto

File: src/theory/arith/nonlinear_extension.cpp [hunk 3]
```diff
@@ -3360,6 +3360,53 @@ std::vector<Node> NonlinearExtension::checkTangentPlanes() {
                       << "Tangent plane lemma : " << tlem << std::endl;
                   lemmas.push_back(tlem);
                 }
+
+                // tangent plane reverse implication
+
+                // t <= tplane -> ( (a <= a_v ^ b >= b_v) v (a >= a_v ^ b <=
+                // b_v) ) in clause form, the above becomes t <= tplane -> a <=
+                // a_v v b <= b_v t <= tplane -> b >= b_v v a >= a_v
+                Node a_leq_av = NodeManager::currentNM()->mkNode(LEQ, a, a_v);
```
Comment (ajreynol) [inline src/theory/arith/nonlinear_extension.cpp]: Can you cache `NodeManager * nm = NodeManager::currentNM();` at the top of this function and use `nm->mkNode` throughout? (This is the new style we've been conforming to; my older code above did not conform to this style, but don't worry about changing it).

**Comment by ahmed-irfan** [comment]
> This looks great, somehow I missed this direction of the implication.
> 
Actually, this was also missing in the papers :)
I discovered this some time ago.

**Comment by ajreynol** [comment]
Would be curious how this does on QF_NRA btw.

**Comment by ahmed-irfan** [comment]
> Would be curious how this does on QF_NRA btw.

Here it is... 60sec timeout... not a big improvement... 
image

**Pull request merged**